### PR TITLE
Feature selection

### PR DESCRIPTION
**Pull request by dan-tee**

### CODE REVIEW

File: guacml/metrics/accuracy.py [hunk 1]
```diff
@@ -6,11 +6,11 @@
 class Accuracy(BaseEvalMetric):
     @staticmethod
     def row_wise_error(truth, prediction):
-        return -Accuracy.correct(truth, prediction)
+        return -100 *Accuracy.correct(truth, prediction)
 
     @staticmethod
     def error(truth, prediction):
-        return -Accuracy.correct(truth, prediction).sum() / truth.shape[0]
+        return -100 * Accuracy.correct(truth, prediction).sum() / truth.shape[0]
```
Comment (hwartig) [inline guacml/metrics/accuracy.py]: Not so sure about this one. It seems to me this is only for displaying it nicely to the user, right? I don't feel that this is the right place however to deal with user interaction as `Accuracy` currently may also be used as eval metric. If we want to format Accuracy as percent for the user I would do it elsewhere and also append the '%' sign while at it. WDYT?

File: guacml/models/base_model.py [hunk 2]
```diff
@@ -18,8 +18,15 @@ def train(self, x, y, **kwargs):
...
     def to_int(value):
         if value is None:
             return None
-        return int(value)
\ No newline at end of file
+        return int(value)
+
+    @staticmethod
+    def pos_int(value):
+        return max(int(value), 0)
```
Comment (hwartig) [inline guacml/models/base_model.py]: Did you know about the [convention](https://robots.thoughtbot.com/no-newline-at-end-of-file) that every text file (incl source code) should end with a newline? If you change your editor to automatically add them this weird git errors 'no newline at end of file' will go away.

feature_selection_by_dan-tee_ _pull_request__5_ _guacml_guacml

File: guacml/models/hyper_param_info.py [hunk 3]
```diff
@@ -1,4 +0,0 @@
-class HyperParameterInfo():
```
Comment (hwartig) [inline guacml/models/hyper_param_info.py:1]: 👍  for removal (no need for extra untyped datastructure)

File: guacml/step_tree/feature_reducer.py [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+                if len(reduced_feats) == 1:
+                    break
+        return features
+
+    def reduce_one_by_one(self, features, cv_bs_errors):
+        while len(features) > 1:
+            for to_drop in features:
+                reduced_feats = list(features)
+                reduced_feats.remove(to_drop)
+                #print('Try dropping feature {0}'.format(to_drop))
```
Comment (hwartig) [inline guacml/step_tree/feature_reducer.py]: ✂️

File: guacml/step_tree/feature_reducer.py [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+        return features
+
+    def reduce_using_feature_importance(self, features, cv_bs_errors):
+        max_iterations = 10
+        for i in range(max_iterations):
+            feat_imp = self.model_runner.feature_importances()
+            feat_imp = feat_imp[features].sort_values()
+            cumsums = feat_imp.cumsum() / feat_imp.sum()
+            reduced_feats = cumsums[cumsums <= 0.95].index
+            #print('try dropping features {0}'.format(cumsums[cumsums > 0.95].index.tolist()))
```
Comment (hwartig) [inline guacml/step_tree/feature_reducer.py]: ✂️

File: guacml/models/xgboost.py [hunk 6]
```diff
@@ -1,25 +1,25 @@
...
 from guacml.enums import ProblemType
 from guacml.models.base_model import BaseModel
-from guacml.models.hyper_param_info import HyperParameterInfo
 from guacml.preprocessing.column_analyzer import ColType
 from hyperopt import hp
 
 
-
 class XgBoost(BaseModel):
+
```
Comment (hwartig) [inline guacml/models/xgboost.py:13]: Was the extra line here intentional? I noticed you don't usually put one after a class declaration.

File: demos/titanic/demo.ipynb [hunk 7]
```diff
@@ -271,45 +285,58 @@
...
        "</div>"
       ],
       "text/plain": [
-       "      model name holdout error cv error training error\n",
-       "0       xg_boost       -0.8101  -0.8042        -0.8822\n",
-       "1  random_forest       -0.8101  -0.8462        -0.7698\n",
-       "2   linear_model       -0.7765  -0.8671        -0.8014"
+       "      model name  n features holdout error holdout error interval cv error  \\\n",
+       "1  random_forest           9        -82.68         -86.59, -79.89   -79.72   \n",
+       "0       xg_boost           1        -79.89          -83.8, -75.98   -76.92   \n",
```
Comment (hwartig) [inline demos/titanic/demo.ipynb]: I'm confused by this one:

model | name | n features | holdout error | holdout error interval | cv error
------ | ------ | ---------- | ------------- | --------------------- | --------
0 | xg_boost | **=> 1 <=** | -79.89 | -83.8, -75.98 | -76.92

Does that imply that the feature selector removed every but one feature?

File: guacml/metrics/accuracy.py (hunk 1 above)
Comment (dan-tee) [reply guacml/metrics/accuracy.py]: Totally agree.

File: guacml/models/base_model.py (hunk 2 above)
Comment (dan-tee) [reply guacml/models/base_model.py]: Done.

File: demos/titanic/demo.ipynb (hunk 7 above)
Comment (dan-tee) [reply demos/titanic/demo.ipynb]: This is possible. Things don't work very smoothly on the Titanic data set yet. It has very few rows, so there is lots of uncertainty in all calculation. Hope 3-fold cross validation will make this better.

Comment (dan-tee) [reply demos/titanic/demo.ipynb]: You also see this in the holdout error interval, which is huge.

File: guacml/step_tree/feature_reducer.py (hunk 5 above)
Comment (dan-tee) [reply guacml/step_tree/feature_reducer.py]: What do you mean by the scissors?

File: demos/titanic/demo.ipynb (hunk 7 above)
Comment (hwartig) [reply demos/titanic/demo.ipynb]: ok makes sense

File: guacml/step_tree/feature_reducer.py (hunk 5 above)
Comment (hwartig) [reply guacml/step_tree/feature_reducer.py]: I would cut that line ;-)

Comment (hwartig) [reply guacml/step_tree/feature_reducer.py]: It looked like a print debug that you might have missed to remove.

File: guacml/step_tree/model_manager.py [hunk 8]
```diff
@@ -0,0 +1,55 @@
...
+
+    def execute(self, input, metadata):
+        model_runner = ModelRunner(self.model, input, self.target, self.eval_metric, self.splitter)
+        features = self.model.select_features(metadata)
+        features = features[features != self.target]
+
+        hp_optimizer = HyperParameterOptimizer(model_runner, features)
+        all_trials, best_hps = hp_optimizer.optimize(self.hyper_param_iterations)
+
+        feature_reducer = FeatureReducer(model_runner, best_hps)
```
Comment (hwartig) [inline guacml/step_tree/model_manager.py:26]: I wonder how much the `best_hps` might change if we remove features. My guess would be that they might change significantly (depending on the model type) and then we should rerun the `HyperParameterOptimiser` for each removed feature again. ... But this is probably for a future revision as we might be able to find a heuristic that helps us deciding when to rerun the `HyperParameterOptimizer`

File: guacml/step_tree/model_runner.py [hunk 9]
```diff
@@ -1,45 +1,73 @@
...
-                           best,
-                           all_trials)
-
-    def score_model(self, input, features):
-        predictions = self.model.predict(input[features])
-        return self.eval_metric.error(input[self.target], predictions), predictions
+    def train_model(self, features, hyper_params):
+        self.features = features
+        self.hyper_params = hyper_params
+        self.model.train(self.train[features], self.train[self.target], **hyper_params)
```
Comment (hwartig) [inline guacml/step_tree/model_runner.py]: This might be unrelated to this PR but I'm wondering why we don't also pass the eval metric to the model train step. Shouldn't we somehow tell the model that it should use our `eval_metric`? What am I missing?

File: guacml/step_tree/model_manager.py [hunk 10]
```diff
@@ -0,0 +1,55 @@
...
+    def __init__(self, model, target, hyper_param_iterations, eval_metric):
+        self.model = model
+        self.target = target
+        self.hyper_param_iterations = hyper_param_iterations
+        self.eval_metric = eval_metric
+        self.splitter = None
+
+    def execute(self, input, metadata):
+        model_runner = ModelRunner(self.model, input, self.target, self.eval_metric, self.splitter)
+        features = self.model.select_features(metadata)
```
Comment (hwartig) [inline guacml/step_tree/model_manager.py]: Did you consider moving the `select_features(metadata)` method from the Model into the ModelManager? It seems to be only used here and it could from the looks of it also be a static method.

We could however go the exactly opposite direction and initialize the Model already with the features and target etc. Then the ModelManager could also get the target and the eval_metric from the Model. Further the model would then already encapsulate all the info it needs to be trained. In addition we could make it immutable and when we want to remove a feature we clone it with the metadata changed for the new instance. This could come in handy when we want to diff models between steps. It would be still able to reference the same dataset as only metadata would have to be changed from one Model revision to another. I would however postpone this bigger refactoring into a separate PR as I'm also still a bit unsure about the downsides of this approach scikit-learn also goes a different direction with their model classes so it might be worth considering staying somewhat consistent with their api choices.

File: guacml/step_tree/model_manager.py [hunk 11]
```diff
@@ -0,0 +1,55 @@
...
+
+        model_runner.train_model(features, best_hps)
+        training_error = model_runner.training_error()
+        holdout_predictions = model_runner.holdout_predictions()
+        holdout_error = model_runner.holdout_error()
+        holdout_error_interval = model_runner.holdout_error_interval()
+        holdout_row_errors = model_runner.row_wise_holdout_error()
+
+        holdout = model_runner.holdout.copy()
+        holdout['prediction'] = holdout_predictions
```
Comment (hwartig) [inline guacml/step_tree/model_manager.py:46]: There are some cases where we assign a plural name to a singular name or vice versa. WDYT about using the singular error if we talk about a single number and plural errors if we talk about a list, vector or dataset?

File: guacml/step_tree/model_manager.py (hunk 8 above)
Comment (dan-tee) [reply guacml/step_tree/model_manager.py:26]: If we move all this to the tree, then it could also be decided by the user. The `HyperParameterOptimiser` is very expensive to run. It needs to retrain the model many times. That's why I just run in once before feature reduction and once after.

File: guacml/step_tree/model_manager.py (hunk 11 above)
Comment (dan-tee) [reply guacml/step_tree/model_manager.py:46]: Sounds good to me.

File: guacml/step_tree/model_runner.py (hunk 9 above)
Comment (dan-tee) [reply guacml/step_tree/model_runner.py]: Many models support few error metrics. Mostly least squared for regression and logloss for classification.

We could later have a mapping where we map the eval metric to the best fitting training metric per model.

**Comment by dan-tee** [comment]
Just added k-fold cross validation.
